### PR TITLE
feat: [CDS-47364]: optimising logic for reconciliation

### PR DESCRIPTION
**Pull request by abhinavmittal-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42000)

### CODE REVIEW

File: 400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceHelper.java [hunk 1]
```diff
@@ -258,6 +192,89 @@ public static ReconciliationStatus performReconciliationHelper(String accountId,
...
+          accountId, new Date(durationStartTs), new Date(durationEndTs));
+    }
+
+    if (featureFlagService.isEnabled(FeatureName.DEPLOYMENT_RECONCILIATION_LOGIC_QUERY_OPTIMIZATIONS, accountId)) {
+      statusMismatchDetected =
+          executionEntity.getReconService().isStatusMismatchedAndUpdatedV2(accountId, durationStartTs, durationEndTs,
+              executionEntity.getSourceEntityClass().getCanonicalName(), executionEntity.getCompletedExecutionsQuery());
+    } else {
+      Map<String, String> tsdbRunningWFs = getRunningWFsFromTSDB(accountId, durationStartTs, durationEndTs,
+          timeScaleDBService, executionEntity.getRunningExecutionQuery(), sourceEntityClass);
```
Comment (abhinavmittal-harness) [inline 400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceHelper.java:234]: This is the only modification method for logic optimisation

File: 400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java [hunk 2]
```diff
@@ -125,15 +126,75 @@ public boolean isStatusMismatchedAndUpdated(Map<String, String> tsdbRunningWFs)
...
+                                         .in(ExecutionStatus.finalStatuses())
+                                         .project(WorkflowExecutionKeys.serviceExecutionSummaries, false);
+    long totalCompletedExecutionsInPrimary = query.count();
+    long totalCompletedExecutionsInSecondary = getCompletedExecutionsFromTSDB(
+        accountId, durationStartTs, durationEndTs, timeScaleDBService, sourceEntityClass, completedExecutionsQuery);
+
+    if (totalCompletedExecutionsInSecondary != totalCompletedExecutionsInPrimary) {
+      try (HIterator<WorkflowExecution> iterator = new HIterator<>(query.fetch(options))) {
+        for (WorkflowExecution workflowExecution : iterator) {
+          updateTimeScaleDbStatuses(workflowExecution, timeScaleDBService);
```
Comment (abhinavmittal-harness) [inline 400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java:158]: Later we can move the saving to timescale to bulk later

**Comment by sahilhindwani-harness** [comment]
*[`400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java` line 141 at r3](https://reviewable.io/reviews/harness/harness-core/42000#-NKRJE8OA5Wvfh_3zYnr:-NKRJE8PE-sr9x9aErlc:b41e3ph) ([raw file](https://github.com/harness/harness-core/blob/e07a5cd23e4228134e78475fdc8830c6969ab426/400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java#L141)):*
> ```Java
>       String sourceEntityClass, String completedExecutionsQuery) {
>     FindOptions options = persistence.analyticNodePreferenceOptions();
>     Query<WorkflowExecution> query = persistence.createQuery(WorkflowExecution.class, excludeAuthority)
> ```

Can we add projections in this query when we are fetching?Maybe,  get only status and executionId?

**Comment by sahilhindwani-harness** [comment]
*[`400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java` line 151 at r3](https://reviewable.io/reviews/harness/harness-core/42000#-NKRJbnM-a7kDOKiEWMl:-NKRJbnM-a7kDOKiEWMm:b-xahdte) ([raw file](https://github.com/harness/harness-core/blob/e07a5cd23e4228134e78475fdc8830c6969ab426/400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java#L151)):*
> ```Java
>                                          .in(ExecutionStatus.finalStatuses())
>                                          .project(WorkflowExecutionKeys.serviceExecutionSummaries, false);
>     long totalCompletedExecutionsInPrimary = query.count();
> ```

Not sure if a good idea or not , can we do updates in timescaleDb based on status rather than executionId. We know that status will be limited and if we can construct a map of status to execution ids from mongo then we can fire update query for those executionIds. 
This would limit the number of operations on timescale.

**Comment by abhinavmittal-harness** [comment]
*[`400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java` line 141 at r3](https://reviewable.io/reviews/harness/harness-core/42000#-NKRJE8OA5Wvfh_3zYnr:-NKRWHawA32O45Swx2yc:b-a321cz) ([raw file](https://github.com/harness/harness-core/blob/e07a5cd23e4228134e78475fdc8830c6969ab426/400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java#L141)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

Can we add projections in this query when we are fetching?Maybe,  get only status and executionId?
</blockquote></details>

We can do in future.
Right now I have added a log that at the time mismatch logic query is hit and by any chance that entity doesn't even exist. so in that scenario I want to update the entire collection.
If we don't see any such logs in the next month we can update this to just update status value and endTime in timescale.
Will add a todo nevertheless

**Comment by abhinavmittal-harness** [comment]
*[`400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java` line 151 at r3](https://reviewable.io/reviews/harness/harness-core/42000#-NKRJbnM-a7kDOKiEWMl:-NKRWqdO3zOO_bhSIgQh:b-limpuc) ([raw file](https://github.com/harness/harness-core/blob/e07a5cd23e4228134e78475fdc8830c6969ab426/400-rest/src/main/java/io/harness/event/reconciliation/service/DeploymentReconServiceImpl.java#L151)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

Not sure if a good idea or not , can we do updates in timescaleDb based on status rather than executionId. We know that status will be limited and if we can construct a map of status to execution ids from mongo then we can fire update query for those executionIds. 
This would limit the number of operations on timescale.
</blockquote></details>

I really like this idea.
Adding a to do for this.
For now I want to log every entity it is doing a mismatch for and want to know how many entities it ends up doing reconciliation for.
So based on those metrics we can make these changes in the next cut

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 4 of 11 files at r1, 4 of 9 files at r2, 7 of 8 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42000)* status: 16 file(s) to review. 1 of 2 approvals obtained.

**Review by Y0709 — APPROVED** [review]
> Reviewed 3 of 11 files at r1, 3 of 9 files at r2, 4 of 8 files at r5, 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42000)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Comment by abhinavmittal-harness** [comment]
trigger ti4

**Review by Y0709 — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42000)* status: :shipit: complete! ✓

**Pull request merged**